crates/fs/private/node.rs

```diff
@@ -70,7 +71,77 @@ impl PrivateNode {
                 dir.metadata.upsert_mtime(time);
                 Self::Dir(Rc::new(dir))
             }
-        }
+        };
+    }
+
+    /// Generates two random set of bytes.
+    pub fn generate_double_random<R: Rng>(rng: &mut R) -> (HashOutput, HashOutput) {
+        const _DOUBLE_SIZE: usize = HASH_BYTE_SIZE * 2;
+        let [first, second] = unsafe {
+            std::mem::transmute::<[u8; _DOUBLE_SIZE], [[u8; HASH_BYTE_SIZE]; 2]>(
+                rng.random_bytes::<_DOUBLE_SIZE>(),
+            )
+        };
+        (first, second)
+    }
+
+    /// Updates bare name ancestry of private sub tree.
+    #[async_recursion(?Send)]
+    pub(crate) async fn update_ancestry<B: BlockStore, R: Rng>(
+        &mut self,
+        parent_bare_name: Namefilter,
+        hamt: Rc<PrivateForest>,
+        store: &mut B,
+        rng: &mut R,
+    ) -> Result<Rc<PrivateForest>> {
+        let hamt = match self {
+            Self::File(file) => {
+                let mut file = (**file).clone();
+
+                file.header.update_bare_name(parent_bare_name);
+                file.header.reset_ratchet(rng);
+
+                *self = Self::File(Rc::new(file));
+
+                hamt
+            }
+            Self::Dir(old_dir) => {
+                let mut dir = (**old_dir).clone();
+
+                let mut working_hamt = Rc::clone(&hamt);
+                for (name, private_ref) in &old_dir.entries {
+                    let mut node = hamt
+                        .get(private_ref, store)
+                        .await?
+                        .ok_or(FsError::NotFound)?;
+
+                    working_hamt = node
+                        .update_ancestry(dir.header.bare_name.clone(), working_hamt, store, rng)
+                        .await?;
+
+                    dir.entries
+                        .insert(name.clone(), node.get_header().get_private_ref()?);
+                }
+
+                dir.header.update_bare_name(parent_bare_name);
+                dir.header.reset_ratchet(rng);
```

Comment:
Why do we reset the ratchet when we move?

I guess it makes sense for `cp`, but may not be the right thing when we're doing a `mv`. :thinking: 

Ok, I guess it's fine because if Alice `mv`s a file that Bob has access to, then (eventually we'll have a spec for that &) write a tombstone at the old position saying that we've moved the file to the new place & what ratchet it now has.

Reply:
Ah right. Great attention to the detail here :) You're right.